### ISSUE TITLE
Gemma 4 26B-A4B 8-bit: broadcast crash on first generate when repetitionPenalty is set

### ISSUE THREAD

**Issue by angelsbrood** [ml-explore/mlx-swift-lm#258]
## Summary

Inference on `mlx-community/gemma-4-26b-a4b-it-8bit` crashes with `[broadcast_shapes] Shapes (N) and (M) cannot be broadcast.` on the first `generate(...)` call whenever `GenerateParameters.repetitionPenalty != nil`. Setting `repetitionPenalty = nil` (with all other parameters unchanged) lets the same call complete successfully.

The shape `(N)` matches `repetitionContextSize` exactly. The shape `(M)` tracks the input sequence length 1:1.

The same on-disk weights generate cleanly under Python `mlx-vlm 0.4.3` via `mlx_vlm.chat`, so the model files and underlying Metal kernels are not at fault.

The same `GenerateParameters` (including the repetition penalty) work cleanly against Qwen 3.6 35B-A3B 8-bit through `LLMModelFactory` on the same mlx-swift-lm 3.31.3 substrate, so the issue is Gemma 4-specific.

## Reproducer

Standalone SwiftPM package: https://github.com/angelsbrood/Gemma4SwiftRepro

The minimum reproducer is ~25 lines. The bisect that isolated `repetitionPenalty` to the trigger:

```swift
import Foundation
import MLX
import MLXLMCommon
import MLXVLM
import MLXHuggingFace
import Tokenizers

let modelDirectory = URL(filePath: "/path/to/gemma-4-26b-a4b-it-8bit/snapshots/1382fb7268fa62c07b83ffa174ff6fa455fb0686")

let tokenizerLoader = #huggingFaceTokenizerLoader()
let container = try await VLMModelFactory.shared.loadContainer(
    from: modelDirectory, using: tokenizerLoader)

// Trigger: repetitionPenalty != nil. Removing this one parameter
// makes the same call complete successfully.
let parameters = GenerateParameters(
    maxTokens: 32,
    repetitionPenalty: 1.1,
    repetitionContextSize: 64
)
let chat: [Chat.Message] = [.user("Say hello.")]

let stream = try await container.perform { context in
    let userInput = UserInput(chat: chat)
    let lmInput = try await context.processor.prepare(input: userInput)
    return try MLXLMCommon.generate(
        input: lmInput, parameters: parameters, context: context)
}

for await event in stream {
    if case .chunk(let text) = event { print(text, terminator: "") }
}
// Crashes during the first iteration of the stream:
// MLX/ErrorHandler.swift:343: Fatal error: [broadcast_shapes] Shapes (64) and (80) cannot be broadcast.
//   at .../mlx-swift/Source/Cmlx/mlx-c/mlx-c/ops.cpp:3952
```

### Bisect summary

Tested against `mlx-community/gemma-4-26b-a4b-it-8bit` via `VLMModelFactory.shared`:

| Variant | repetitionPenalty | Result |
|---|---|---|
| Bare `[.user("Say hello.")]`, default params | nil (default) | ✅ generates `"Hello! How can I help you today?"` |
| + ~1.5K-char system prompt | nil | ✅ |
| + tools array via `UserInput(chat:tools:)` | nil | ✅ generates Gemma 4 native `<\|tool_call>` text |
| + full real-world params (temp 0.7, topP 0.95, **repPenalty 1.1**, repContextSize 64, maxTokens 4096) | 1.1 | ❌ `Shapes (64) and (1139) cannot be broadcast` |
| Bare prompt + **repPenalty 1.1, repContextSize 64 only** (no system, no tools, no temp/topP change) | 1.1 | ❌ `Shapes (64) and (80) cannot be broadcast` |

Cross-substrate verification:

| Stack | Same model + similar params | Result |
|---|---|---|
| Python `mlx_vlm.chat` 0.4.3, same on-disk weights | n/a (default sampling) | ✅ |
| mlx-swift-lm 3.31.3 + Qwen 3.6 35B-A3B 8-bit via `LLMModelFactory`, identical `GenerateParameters` | repPenalty 1.1, repContextSize 64 | ✅ |

## Suspected location

The fact that `(N)` always equals `repetitionContextSize` and `(M)` always equals the input sequence length suggests the broadcast failure is in the rep-penalty application path — likely `Libraries/MLXLMCommon/Evaluate.swift::RepetitionContext.process(logits:)` or similar, where the `(repetitionContextSize,)`-shaped index ring is gathered against the model's logits. Gemma 4's logits shape (or an intermediate exposed when rep penalty is on) doesn't broadcast against the rep-penalty index buffer the way Qwen 3.x's does.

Haven't traced into MLXLMCommon to confirm — happy to dig further if useful.

## Environment

- **mlx-swift-lm:** `3.31.3` (resolved revision: `1c05248bb0899e2a7a4962b84d319cf12f4e12aa`)
- **mlx-swift:** `0.31.3` (resolved revision: `61b9e011e09a62b489f6bd647958f1555bdf2896`)
- **swift-transformers:** `1.3.0` (`Transformers` product, added as direct dep — the `#huggingFaceTokenizerLoader()` macro requires it and v3 doesn't pull it transitively)
- **Swift toolchain:** `Apple Swift version 6.3.1 (swiftlang-6.3.1.1.2 clang-2100.0.123.102)`, target `arm64-apple-macosx26.0`
- **Xcode:** `26.4.1`
- **macOS:** `26.4.1`
- **Hardware:** Apple M5 Pro, 18-core CPU, 64 GB RAM
- **Model:** `mlx-community/gemma-4-26b-a4b-it-8bit`, snapshot `1382fb7268fa62c07b83ffa174ff6fa455fb0686`

## Notes

- Model weights pass an MD5 sanity check against the HF source, and load + generate cleanly via Python `mlx_vlm 0.4.3`.
- Haven't tested the 4-bit, BF16, or other Gemma 4 quantization variants for this same trigger; the bisect of `repetitionPenalty` against the 8-bit was sufficient for my downstream use case (a default change in my app — `repetitionPenalty: nil` is now the substrate default, since for instruction-tuned models in 2026 it's largely vestigial). If it would help triage, happy to run the same bisect against the 4-bit variant.
- The standalone repro package's full README (with experiment outcomes A→D + bisect E) is at https://github.com/angelsbrood/Gemma4SwiftRepro if you want the longer form.
